pkg/server/cloud/aws/network.go

```diff
@@ -316,57 +316,85 @@ func (e *AwsEC2) ModifySourceDestinationCheck(instanceID string, isEnabled bool)
+		return fmt.Errorf(
+			"invalid input: at least one non-empty value needed (got %q %q)",
+			destinationCIDR, instanceID)
+	}
 	out, err := e.client.DescribeRouteTables(&ec2.DescribeRouteTablesInput{
 		Filters: []*ec2.Filter{
 			{
-				Name: aws.String("vpc-id"),
-				Values: []*string{
-					aws.String(e.vpcID),
-				},
+				Name:   aws.String("vpc-id"),
+				Values: aws.StringSlice([]string{e.vpcID}),
+			},
+			{
+				Name:   aws.String("association.main"),
+				Values: aws.StringSlice([]string{"false"}),
+			},
+			{
+				Name:   aws.String("association.subnet-id"),
+				Values: aws.StringSlice([]string{e.subnetID}),
 			},
 		},
 	})
 	if err != nil {
-		return util.WrapError(err, "Could not list vpc route tables")
+		return util.WrapError(err, "listing vpc route tables")
 	}
 	for _, table := range out.RouteTables {
 		if len(table.Associations) == 0 {
 			continue
 		}
 		for _, route := range table.Routes {
-			if destinationCIDR == aws.StringValue(route.DestinationCidrBlock) {
-				_, err = e.client.DeleteRoute(&ec2.DeleteRouteInput{
-					DestinationCidrBlock: route.DestinationCidrBlock,
-					RouteTableId:         table.RouteTableId,
-				})
-				if err != nil {
-					return util.WrapError(
-						err,
-						"Error deleting old route table entry for table %s",
-						*table.RouteTableId,
-					)
-				}
-				break
+			if destinationCIDR != "" &&
+				destinationCIDR != aws.StringValue(route.DestinationCidrBlock) {
+				continue
+			}
+			// Ignore any routes that are not instance routes.
+			routeIID := aws.StringValue(route.InstanceId)
+			if routeIID == "" || instanceID != "" && instanceID != routeIID {
+				continue
 			}
+			_, err = e.client.DeleteRoute(&ec2.DeleteRouteInput{
+				DestinationCidrBlock: route.DestinationCidrBlock,
+				RouteTableId:         table.RouteTableId,
+			})
+			if err != nil {
+				return util.WrapError(
+					err,
+					"deleting route %s in table %s",
+					destinationCIDR, *table.RouteTableId)
+			}
+			break
 		}
 	}
 	return nil
 }
 
 func (e *AwsEC2) AddRoute(destinationCIDR, instanceID string) error {
+	if destinationCIDR == "" || instanceID == "" {
+		return fmt.Errorf(
+			"invalid input: empty value (got %q %q)",
+			destinationCIDR, instanceID)
+	}
 	out, err := e.client.DescribeRouteTables(&ec2.DescribeRouteTablesInput{
 		Filters: []*ec2.Filter{
 			{
-				Name: aws.String("vpc-id"),
-				Values: []*string{
-					aws.String(e.vpcID),
-				},
+				Name:   aws.String("vpc-id"),
+				Values: aws.StringSlice([]string{e.vpcID}),
+			},
+			{
+				Name:   aws.String("association.main"),
+				Values: aws.StringSlice([]string{"false"}),
+			},
+			{
+				Name:   aws.String("association.subnet-id"),
+				Values: aws.StringSlice([]string{e.subnetID}),
 			},
 		},
 	})
 	if err != nil {
-		return util.WrapError(err, "Could not list vpc route tables")
+		return util.WrapError(err, "listing vpc route tables")
 	}
 	for _, table := range out.RouteTables {
```

Comment:
Any idea what happens if we try to add 2 identical destinationCIDRs to a route table?  If that causes an error, one thing we could do is catch that error, delete the old route and then try to re-add the route.

If it doesn't cause an error: :tada:

Reply:
I think this is answered in a comment later on.  I'm curious how we should handle this. Can we delete the old route?  The alternative is to keep churning instances for this pod.

Reply:
I'd rather not delete an existing route. If the operator accidentally tries to add the same prefix via multiple pods, one will succeed, the others will fail. The cause of the failure will pop up via logs and the pod status. If we silently remove existing active routes, then things will *seem* to work (no imminent pod failure), except likely routing will be misconfigured.

Reply:
BTW this is why I added extra calls to RemoveRoute(): if a pod is terminated, we should ensure that eventually its cloud route is also gone.